src/ring_pedersen.rs

```diff
@@ -0,0 +1,318 @@
+    /// alongside a zero knowledge proof [`PiPrmProof`]
+    /// that the produced commitment scheme is validly constructed.
+    pub(crate) fn extract(
+        sk: &DecryptionKey,
+        rng: &mut (impl RngCore + CryptoRng),
+    ) -> Result<Self> {
+        let ek = sk.encryption_key();
+        let modulus = ek.n();
+        let phi_n = sk.totient();
+        let tau = random_positive_bn(rng, modulus);
+        let lambda = random_positive_bn(rng, phi_n);
+        let t = tau.modpow(&BigNumber::from(2), modulus);
+        let s = t.modpow(&lambda, modulus);
+        let scheme = RingPedersen {
+            modulus: modulus.clone(),
+            s,
+            t,
+        };
+        let proof = PiPrmProof::prove(
+            rng,
+            &PiPrmInput::new(scheme.clone()),
+            &PiPrmSecret::new(&lambda, phi_n),
+        )?;
+        Ok(Self { scheme, proof })
+    }
+
+    /// Verifies that the underlying [`RingPedersen`] commitment scheme was constructed correctly
+    /// according to the associated [`PiPrmProof`].
+    pub(crate) fn verify(&self) -> Result<()> {
+        self.proof.verify(&PiPrmInput::new(self.scheme.clone()))
+    }
+
+    /// Returns the underlying [`RingPedersen`] commitment scheme associated
+    /// with this object.
+    pub(crate) fn scheme(&self) -> &RingPedersen {
+        &self.scheme
+    }
+
+    /// Generates a [`VerifiedRingPedersen`] object from a random number generator for testing purposes.
+    #[cfg(test)]
+    pub(crate) fn gen(rng: &mut (impl RngCore + CryptoRng)) -> Result<Self> {
+        let (sk, _, _) = DecryptionKey::new(rng)?;
+        Self::extract(&sk, rng)
+    }
+}
+
+impl RingPedersen {
+    /// Returns the underlying modulus.
+    pub(crate) fn modulus(&self) -> &BigNumber {
+        &self.modulus
+    }
+
+    /// Returns the underlying `s` parameter.
+    pub(crate) fn s(&self) -> &BigNumber {
+        &self.s
+    }
+
+    /// Returns the underlying `t` parameter.
+    pub(crate) fn t(&self) -> &BigNumber {
+        &self.t
+    }
+
+    /// Constructs a [`RingPedersen`] object from its constituent parts for testing purposes.
+    #[cfg(test)]
+    pub(crate) fn from_parts(s: BigNumber, t: BigNumber, modulus: BigNumber) -> Self {
+        Self { modulus, s, t }
+    }
+
+    /// Produces commitment randomness sampled from `+/- 2^range * modulus * N`,
+    /// where `N` is the modulus of the commitment scheme.
+    pub(crate) fn commitment_randomness(
+        &self,
+        range: usize,
+        modulus: &BigNumber,
+        rng: &mut (impl RngCore + CryptoRng),
+    ) -> CommitmentRandomness {
+        let randomness = random_plusminus_scaled(rng, range, &(modulus * &self.modulus));
+        CommitmentRandomness(randomness)
+    }
+
+    /// Produces a commitment to `value`, which must be drawn
+    /// from `+/- 2^range`. The commitment randomness is sampled from
+    /// `+/- 2^range * N`, where `N` is the modulus of the commitment scheme.
```

Comment:
nit: 
- Update the wording to be more explicit about what is and isn't enforced. E.g. the expected usage is that `value` is in range, but we don't check (should we check?).
- Per other comment, add the formula for commitments.
- Adjust phrasing to have a one-liner method description, then newline, then all these other details.
```
/// Produces a commitment to `value`.
///
/// blah blah in range
```

Reply:
> should we check?

For robustness it's probably a good idea that we check I think. It might hurt performance though, so there's a tradeoff. Any thoughts?

Reply:
I added a test for this, expecting the commitment to not reconstruct correctly, but it does. Which actually makes sense, there's nothing in the commitment process that restricts `value`'s range.

This suggests to me it makes even more sense to add a check for misuse resistance.

Reply:
Hmm, this is trickier than I thought. `reconstruct` also takes a `value`, but we can't restrict that `value`'s range because it doesn't take `range` as an argument. So is it worth restricting `value`'s range in `commit`, particularly if there's no a priori reason to have that restriction (that is, `commit` works fine over any `value`, not only `value`s in the provided range)?

Reply:
I take it all back (I think): The restriction on `value` causes the proofs to fail :-(. The reason is in the `pifac` step of constructing `R = s^{N_0} t^\sigma`, `N_0` is outside the expected range... To me the cleanest approach would be to remove the restriction on value, but I'm open to other ideas.

Reply:
Hm. Some tension between general-purpose Ring-Pedersen and application-specific commitments we're dealing with here. Let's remove the restriction on `value`. The restrictions vary by use case (even within our application) so I think it's reasonable to push the burden of maintaining that correctness to the proofs that call this API. It's 3 proofs with restrictions and 1 without, which makes it a bit annoying to generalize the API to the one. But we both already audited all the places where we call this to make sure it's correct.